File: Refactor/Speculum.Api/Presentation/Sessions/DataStreamMux.cs
using System.Buffers.Binary;

namespace Speculum.Api.Presentation.Sessions;

/// <summary>
/// Binary mux for one WebSocket data-stream session.
/// OPEN/DATA/CLOSE carry a streamId; pipe bytes (PipeKind + MessagePack frames) are opaque DATA payloads.
/// </summary>
internal static class DataStreamMux
{
    public const byte OpOpen = 1;
    public const byte OpData = 2;
    public const byte OpClose = 3;

    public const int HeaderBytes = 3; // op + streamId u16 BE

    /// <summary>Server-allocated stream ids use the high bit to avoid colliding with client ids.</summary>
    public const ushort ServerStreamIdBase = 0x8000;

    public static byte[] EncodeOpen(ushort streamId)
    {
        var frame = new byte[HeaderBytes];
        frame[0] = OpOpen;
        BinaryPrimitives.WriteUInt16BigEndian(frame.AsSpan(1), streamId);
        return frame;
    }

    public static byte[] EncodeClose(ushort streamId)
    {
        var frame = new byte[HeaderBytes];
        frame[0] = OpClose;
        BinaryPrimitives.WriteUInt16BigEndian(frame.AsSpan(1), streamId);
        return frame;
    }

    public static byte[] EncodeData(ushort streamId, ReadOnlySpan<byte> payload)
    {
        var frame = new byte[HeaderBytes + payload.Length];
        frame[0] = OpData;
        BinaryPrimitives.WriteUInt16BigEndian(frame.AsSpan(1), streamId);
        payload.CopyTo(frame.AsSpan(HeaderBytes));
        return frame;
    }

    public static bool TryParse(
        ReadOnlySpan<byte> frame,
        out byte op,
        out ushort streamId,
        out ReadOnlySpan<byte> payload)
    {
        op = 0;
        streamId = 0;
        payload = default;
        if (frame.Length < HeaderBytes)
        {
            return false;
        }

        op = frame[0];
        streamId = BinaryPrimitives.ReadUInt16BigEndian(frame.Slice(1, 2));
        payload = frame.Length > HeaderBytes ? frame.Slice(HeaderBytes) : ReadOnlySpan<byte>.Empty;
        return op is OpOpen or OpData or OpClose;
    }
}

File: web/src/lib/speculum/dataStreamMux.ts
/** Binary mux for one WebSocket data-stream session — must match API DataStreamMux. */
export const DataStreamMuxOp = {
  Open: 1,
  Data: 2,
  Close: 3,
} as const

export const DATA_STREAM_MUX_HEADER_BYTES = 3
/** Server-allocated stream ids use the high bit. */
export const DATA_STREAM_MUX_SERVER_ID_BASE = 0x8000

export function encodeMuxOpen(streamId: number): Uint8Array {
  const frame = new Uint8Array(DATA_STREAM_MUX_HEADER_BYTES)
  frame[0] = DataStreamMuxOp.Open
  new DataView(frame.buffer).setUint16(1, streamId & 0xffff, false)
  return frame
}

export function encodeMuxClose(streamId: number): Uint8Array {
  const frame = new Uint8Array(DATA_STREAM_MUX_HEADER_BYTES)
  frame[0] = DataStreamMuxOp.Close
  new DataView(frame.buffer).setUint16(1, streamId & 0xffff, false)
  return frame
}

export function encodeMuxData(streamId: number, payload: Uint8Array): Uint8Array {
  const frame = new Uint8Array(DATA_STREAM_MUX_HEADER_BYTES + payload.byteLength)
  frame[0] = DataStreamMuxOp.Data
  new DataView(frame.buffer).setUint16(1, streamId & 0xffff, false)
  frame.set(payload, DATA_STREAM_MUX_HEADER_BYTES)
  return frame
}

export function tryParseMuxFrame(
  frame: Uint8Array,
): { op: number; streamId: number; payload: Uint8Array } | null {
  if (frame.byteLength < DATA_STREAM_MUX_HEADER_BYTES) {
    return null
  }
  const op = frame[0]!
  if (
    op !== DataStreamMuxOp.Open &&
    op !== DataStreamMuxOp.Data &&
    op !== DataStreamMuxOp.Close
  ) {
    return null
  }
  const streamId = new DataView(frame.buffer, frame.byteOffset, frame.byteLength).getUint16(1, false)
  const payload =
    frame.byteLength > DATA_STREAM_MUX_HEADER_BYTES
      ? frame.subarray(DATA_STREAM_MUX_HEADER_BYTES)
      : new Uint8Array(0)
  return { op, streamId, payload }
}
